core/src/main/java/org/elasticsearch/action/support/TransportAction.java

```diff
@@ -246,4 +250,37 @@ public void onFailure(Throwable e) {
             listener.onFailure(e);
         }
     }
+
+    /**
+     * Wrapper for an action listener that persists the result at the end of the execution
+     */
+    private static class PersistentActionListener<Response extends ActionResponse> implements ActionListener<Response> {
+        private final ActionListener<Response> listener;
+        private final Task task;
+        private final TaskManager taskManager;
+
+        private  PersistentActionListener(TaskManager taskManager, Task task, ActionListener<Response> listener) {
+            this.taskManager = taskManager;
+            this.task = task;
+            this.listener = listener;
+        }
+
+        @Override
+        public void onResponse(Response response) {
+            try {
+                taskManager.persistResult(task, response, listener);
+            } catch (Throwable e) {
+                listener.onFailure(e);
+            }
+        }
+
+        @Override
+        public void onFailure(Throwable e) {
+            try {
+                taskManager.persistResult(task, e, listener);
+            } catch (Throwable e1) {
+                listener.onFailure(e1);
```

Comment:
s/listener/delegate/? I read this and immediately thought "infinite loop!" because this thing already **is** a listener. I know it is silly though.